src/components/implementation/mem_mgr/parsec/mem_man_parsec.c

```diff
@@ -0,0 +1,1381 @@
+		/* first child case */
+		if (p->child == m) p->child = m->sibling_next;
+		mapping_unlock(p);
+	}
+ 
+	ret = 0;
+done:
+	return ret;
+}
+
+static int 
+mapping_del_children(mapping_t *m, comp_t *c)
+{
+	mapping_t *child;
+	int ret = -EINVAL;
+
+	mapping_lock(m);
+	/* always re-check before modification. */
+	if (!parsec_item_active(m)) goto done;
+
+	/* Free all the children */
+	child = m->child;
+	while (child) {
+		mapping_free(child, c);
+		child = child->sibling_next;
+	}
+
+	ret = 0;
+done:
+	mapping_unlock(m);
+
+	return ret;
+}
+
+int mman_revoke_page(spdid_t compid, vaddr_t addr, int flags)
+{
+	comp_t *c;
+	mapping_t *m;
+	int ret = 0;
+
+	/* printc("in revoke page from comp %d @ addr %x\n", compid, addr); */
+
+	parsec_read_lock(&mm);
+
+	c = comp_lookup(compid);
+	if (unlikely(!c)) goto done;
+	m = mapping_lookup(c, addr);
+	if ((!m)) goto done;
+
+	ret = mapping_del_children(m, c);
+done:	
+	parsec_read_unlock(&mm);
+
+	return 0;
+}
+
+/* This releases the non-head mappings. They won't go back to any
+ * queue / freelist. */
+static inline int 
+release_multipage(mapping_t *m, comp_t *c)
+{
+	int npages, i;
+	vaddr_t page;
+	mapping_t *temp;
+
+	npages = parsec_item_size(m) / PAGE_SIZE;
+	assert(parsec_item_size(m) % PAGE_SIZE == 0);
+
+	page = m->vaddr;
+	/* release non-head mappings only. */
+	for (i = 1; i < npages; i++) {
+		page += PAGE_SIZE;
+		temp = mapping_lookup(c, page);
+		if (temp) mapping_del_all(temp, c);
+	}
+
+	return 0;
+}
+
+int mman_release_page(spdid_t compid, vaddr_t addr, int flags)
+{
+	comp_t *c;
+	mapping_t *m;
+	int ret = 0;
+
+	/* printc("in release page from comp %d @ addr %x\n", compid, addr); */
+
+	parsec_read_lock(&mm);
+
+	c = comp_lookup(compid);
+	if (unlikely(!c)) goto done;
+	m = mapping_lookup(c, addr);
+	if ((!m)) goto done;
+
+	/* A hack for now. Should solve it better. */
+	if (parsec_item_size(m) > PAGE_SIZE) 
+		release_multipage(m, c);
+
+	ret = mapping_del_all(m, c);
+done:	
+	parsec_read_unlock(&mm);
+
+	return 0;
+}
+
+void mman_print_stats(void) {}
+
+void mman_release_all(void) {}
+
+PERCPU_ATTR(static volatile, int, initialized_core); /* record the cores that still depend on us */
+
+void cos_upcall_fn(upcall_type_t t, void *arg1, void *arg2, void *arg3)
+{
+	printc("upcall in mm, %ld %d\n", cos_cpuid(), cos_get_thd_id());
+	/* printc("cpu %ld: thd %d in mem_mgr init. args %d, %p, %p, %p\n", */
+	/*        cos_cpuid(), cos_get_thd_id(), t, arg1, arg2, arg3); */
+	switch (t) {
+	case COS_UPCALL_THD_CREATE:
+		if (cos_cpuid() == INIT_CORE) {
+			int i;
+			for (i = 0; i < NUM_CPU; i++)
+				*PERCPU_GET_TARGET(initialized_core, i) = 0;
+			mm_init(); 
+		} else {
+			/* Make sure that the initializing core does
+			 * the initialization before any other core
+			 * progresses */
+			while (*PERCPU_GET_TARGET(initialized_core, INIT_CORE) == 0) ;
+		}
+		*PERCPU_GET(initialized_core) = 1;
+		break;			
+	default:
+		BUG(); return;
+	}
+
+	return;
+}
```

Comment:
I'll go through the logic for all of this and parsec later.  I'm doing a bare-minimum code-review now to get this accepted.  I assume that we'll continue to improve it.